trieste/models/optimizer.py

```diff
@@ -0,0 +1,118 @@
+from typing import Any, Dict, Optional, Callable, Union, Iterable, Tuple
+from functools import singledispatch
+from dataclasses import dataclass, field
+import tensorflow as tf
+import gpflow
+
+from gpflow.models import ExternalDataTrainingLossMixin, InternalDataTrainingLossMixin
+from ..utils import jit
+from ..datasets import Dataset
+
+
+DatasetBuilder = Callable[[Dataset, Optional[int]], Union[Iterable, Tuple]]
+"""
+Type alias for a function that creates minibatches from a :class:`~optiflow.datasets.Dataset`.
+"""
+
+
+@dataclass
+class Optimizer:
+    optimizer: Union[gpflow.optimizers.Scipy, tf.optimizers.Optimizer]
+    minimize_args: Dict[str, Any] = field(default_factory=lambda: {})
+    compile: bool = False
+
+    def create_loss(self, model: tf.Module, dataset: Dataset) -> Callable:
+        x = tf.convert_to_tensor(dataset.query_points)
+        y = tf.convert_to_tensor(dataset.observations)
```

Comment:
we aren't using `tf.convert_to_tensor` anywhere else, so do we need it here?

Reply:
btw you can also `from dataclasses import astuple; astuple(dataset)` to unpack a dataset

Reply:
I know about `astuple` and convert_to_tensor here is for safety reasons if query_points or observations are numpys.